python/taichi/lang/matrix.py

```diff
@@ -259,21 +260,60 @@ def _get_slice(self, a, b):
             b = range(b.start or 0, b.stop or self.m, b.step or 1)
         return Matrix([[self(i, j) for j in b] for i in a])
 
+    def _cal_slice(self, index, dim):
+        start, stop, step = index.start or 0, index.stop or (
+            self.n if dim is 0 else self.m), index.step or 1
+
+        def helper(x):
+            if isinstance(x, expr.Expr):
+                if isinstance(x.ptr, int):
+                    return x.ptr
+                raise TaichiSyntaxError(
+                    "The element type of slice of Matrix/Vector index must be a compile-time constant integer!"
+                )
+            return x
+
+        start, stop, step = helper(start), helper(stop), helper(step)
+        return [_ for _ in range(start, stop, step)]
+
     @taichi_scope
     def _subscript(self, *indices):
         assert len(indices) in [1, 2]
         i = indices[0]
         j = 0 if len(indices) == 1 else indices[1]
-        if isinstance(i, slice) or isinstance(j, slice):
-            for a in (i, j):
-                if isinstance(a, slice):
-                    if isinstance(a.start, expr.Expr) or isinstance(
-                            a.step, expr.Expr) or isinstance(
-                                a.stop, expr.Expr):
-                        raise TaichiSyntaxError(
-                            "The element type of slice of Matrix/Vector index must be a compile-time constant integer!"
-                        )
-            return self._get_slice(i, j)
+        has_slice = False
+        if isinstance(i, slice):
+            i = self._cal_slice(i, 0)
+            has_slice = True
+        if isinstance(j, slice):
+            j = self._cal_slice(j, 1)
+            has_slice = True
+
+        if has_slice:
+            if not isinstance(i, list):
+                i = [i]
+            if not isinstance(j, list):
+                j = [j]
+
+            if self.local_tensor_proxy is not None:
+                assert self.dynamic_index_stride is not None
```

Comment:
The logic of this part seems incomplete - you should directly call `subscript` again for single variable index in order to handle the other cases, where the matrix is an element of a field or ndarray.